### PR TITLE
Support ETCD encryption key secret rotation for shoot clusters

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
Similar to #5968, this PR introduces the possibility to rotate the ETCD encryption key secrets for shoot clusters.

Generally, the implementation follows the [official description](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#rotating-a-decryption-key) by the Kubernetes community and is similar to how shoot CA or `ServiceAccount` signing key rotation works:

1. There are four phases: `Preparing`, `Prepared`, `Completing`, `Completed`.
2. In phase `Preparing`, the following things are happening:
  1. A new key is generated and added as second key to the `EncryptionConfiguration` read by `kube-apiserver`.
  2. All `kube-apiserver` replicas are rolled out. After this is completed, all replicas can use the new key for decryption while the old key is still used for encryption (since it's the first key).
  3. Now the new key is made the first key in the `EncryptionConfiguration`.
  4. All `kube-apiserver` replicas are rolled out again. After this is completed, all replicas will use the new key for encryption (since it's the first key now) while the old key is still used for decryption.
  5. Now all `Secret`s are patched with a label containing the name of the current encryption key secret. This will make the `kube-apiserver` rewrite them in ETCD, effectively encrypting the data with the new key.
3. In phase `Completing`, the label is removed again from all `Secret`s.

Two things "good to know":
- Despite not being technically required, we are now patching all `Secret`s twice. The reason is that we use the label to identify which of the `Secret`s have already been patched so that we do not try it again in case the flow has failed and is being retried. This is to not overload the `kube-apiserver` for clusters with many thousands of `Secret`s. We already follow a similar strategy when rotating the `ServiceAccount` secrets.
- In the `Preparing` phase, after we first rolled out the `kube-apiserver`s and populated the new encryption key, we add an annotation to the `kube-apiserver`. This annotation is used to "remember" that we have already successfully populated the new key and can make future deployments using the new key for encryption. This is to prevent that `kube-apiserver`s are rolled out too often in case the shoot reconciliation flow fails somewhen during the `Preparing` phase and is being retried. This annotation is just used as an optimization and technically not required, i.e. there is no problem in case we lose it (for what reason this anyways would happen).

**Release note**:

```feature user
It is now possible to trigger the rotation of the ETCD encryption key secret for shoot clusters. Please consult the [documentation](https://github.com/gardener/gardener/tree/master/docs/usage/shoot_credentials_rotation.md#etcd-encryption-key) for more information.
```

### CODE REVIEW

**Comment by oliver-goetz** [comment]
Tested it locally, thought two times there might be something, but there wasn't
lgtm ...without a slash for now 😄

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 1]
```diff
@@ -102,16 +102,18 @@ type Interface interface {
 	SetAutoscalingAPIServerResources(corev1.ResourceRequirements)
 	// SetAutoscalingReplicas sets the Replicas field in the AutoscalingConfig of the Values of the deployer.
 	SetAutoscalingReplicas(*int32)
+	// SetETCDEncryptionConfig sets the ETCDEncryptionConfig field int he Values of the deployer.
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]:
```suggestion
	// SetETCDEncryptionConfig sets the ETCDEncryptionConfig field in the Values of the deployer.
```

File: pkg/apis/core/helper/helpers_test.go [hunk 2]
```diff
@@ -111,6 +111,18 @@ var _ = Describe("helper", func() {
...
 	)
 
+	DescribeTable("#GetShootETCDEncryptionKeyRotationPhase",
+		func(credentials *core.ShootCredentials, expectedPhase core.ShootCredentialsRotationPhase) {
+			Expect(GetShootETCDEncryptionKeyRotationPhase(credentials)).To(Equal(expectedPhase))
+		},
+
+		Entry("credentials nil", nil, core.ShootCredentialsRotationPhase("")),
+		Entry("rotation nil", &core.ShootCredentials{}, core.ShootCredentialsRotationPhase("")),
+		Entry("ca nil", &core.ShootCredentials{Rotation: &core.ShootCredentialsRotation{}}, core.ShootCredentialsRotationPhase("")),
```
Comment (ialidzhikov) [inline pkg/apis/core/helper/helpers_test.go]:
```suggestion
		Entry("etcdEncryptionKey nil", &core.ShootCredentials{Rotation: &core.ShootCredentialsRotation{}}, core.ShootCredentialsRotationPhase("")),
```

File: pkg/apis/core/v1beta1/helper/helper_test.go [hunk 3]
```diff
@@ -2400,8 +2400,34 @@ var _ = Describe("helper", func() {
...
+	)
+
+	DescribeTable("#GetShootETCDEncryptionKeyRotationPhase",
+		func(credentials *gardencorev1beta1.ShootCredentials, expectedPhase gardencorev1beta1.ShootCredentialsRotationPhase) {
+			Expect(GetShootETCDEncryptionKeyRotationPhase(credentials)).To(Equal(expectedPhase))
+		},
+
+		Entry("credentials nil", nil, gardencorev1beta1.ShootCredentialsRotationPhase("")),
+		Entry("rotation nil", &gardencorev1beta1.ShootCredentials{}, gardencorev1beta1.ShootCredentialsRotationPhase("")),
+		Entry("ca nil", &gardencorev1beta1.ShootCredentials{Rotation: &gardencorev1beta1.ShootCredentialsRotation{}}, gardencorev1beta1.ShootCredentialsRotationPhase("")),
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/helper/helper_test.go]:
```suggestion
		Entry("etcdEncryptionKey nil", &gardencorev1beta1.ShootCredentials{Rotation: &gardencorev1beta1.ShootCredentialsRotation{}}, gardencorev1beta1.ShootCredentialsRotationPhase("")),
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 4]
```diff
@@ -334,12 +334,31 @@ func (b *Botanist) computeKubeAPIServerServerCertificateConfig() kubeapiserver.S
...
+	if gardencorev1beta1helper.GetShootETCDEncryptionKeyRotationPhase(b.Shoot.GetInfo().Status.Credentials) == gardencorev1beta1.RotationPreparing {
+		deployment := &metav1.PartialObjectMetadata{}
+		deployment.SetGroupVersionKind(appsv1.SchemeGroupVersion.WithKind("Deployment"))
+		if err := b.K8sSeedClient.Client().Get(ctx, kutil.Key(b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeAPIServer), deployment); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return kubeapiserver.ETCDEncryptionConfig{}, err
+			}
+		}
+
+		// If the new encryption key was not yet populated to all replicas then we should not still use the old key for
```
Comment (ialidzhikov) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
		// If the new encryption key was not yet populated to all replicas then we should still use the old key for
```

To my understanding if the new encryption key is not yet added to the EncryptionConfiguration, then the old key should be still used for encrypting the data. Right?

File: docs/usage/shoot_credentials_rotation.md [hunk 5]
```diff
@@ -167,8 +167,39 @@ However, this feature gate is deprecated, turned off by default and will be remo
...
+Technically, the `Preparing` phase indicates the stages one and two.
+Once it is completed, the `Prepared` phase indicates readiness for stage three.
+The `Completing` phase indicates stage three, and the `Completed` phase states that the rotation process has finished.
+
+> You can check the `.status.credentials.rotation.etcdEncryptionKey` field in the `Shoot` to see when the rotation was last initiated, last completed, and in which phase it currently is.
+
+In order to start the rotation (stage one), you have to annotate the shoot with the `rotate-etcd-encryption-key-start` operation:
+
+```bash
+kubectl -n <shoot-namespace> annotate shoot <shoot-name> gardener.cloud/operation=rotate-etcd-encryption-key-start
```
Comment (ialidzhikov) [inline docs/usage/shoot_credentials_rotation.md:190]: I executed 2 times the rotation in a row. I noticed that old `kube-apiserver-etcd-encryption-configuration` Secrets are not cleaned up:

```
kube-apiserver-etcd-encryption-configuration-10ebf368             Opaque                                1      108s
kube-apiserver-etcd-encryption-configuration-15b6cf29             Opaque                                1      11m
kube-apiserver-etcd-encryption-configuration-671b6f54             Opaque                                1      5m39s
kube-apiserver-etcd-encryption-configuration-6a3e72c0             Opaque                                1      23h
kube-apiserver-etcd-encryption-configuration-6dc4e190             Opaque                                1      22m
kube-apiserver-etcd-encryption-configuration-9e17b9c8             Opaque                                1      7m1s
kube-apiserver-etcd-encryption-configuration-a884d738             Opaque                                1      24m
```

File: pkg/operation/botanist/kubeapiserver.go (hunk 4 above)
Comment (rfranzke) [reply pkg/operation/botanist/kubeapiserver.go]: Yes, right, thanks!

File: docs/usage/shoot_credentials_rotation.md (hunk 5 above)
Comment (rfranzke) [reply docs/usage/shoot_credentials_rotation.md:190]: Yes, those will only be cleaned up every `12h` by the [garbage collector for immutable `ConfigMap`s/`Secret`s](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#garbage-collector-for-immutable-configmapssecrets): https://github.com/gardener/gardener/blob/af04d906c0a42a6b2dfd5923f21c0c493721ca2b/pkg/operation/botanist/component/resourcemanager/resource_manager.go#L700

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

**Pull request merged**